### PR TITLE
Edit metadata

### PR DESCRIPTION
**Pull request by ManApart**

### CODE REVIEW

File: apps/andi/assets/css/edit.scss [hunk 1]
```diff
@@ -20,27 +20,31 @@
...
-@include area(metadata-form, keywords)
-
-@include area(metadata-form, last-updated)
-@include area(metadata-form, spatial)
-@include area(metadata-form, temporal)
-@include area(metadata-form, organization)
-@include area(metadata-form, level-of-access)
-@include area(metadata-form, language)
-@include area(metadata-form, homepage)
+.div__success-message {
```
Comment (NathanielScottStevens) [inline apps/andi/assets/css/edit.scss]: I wouldn't call this block `div`

File: apps/andi/assets/css/edit.scss [hunk 2]
```diff
@@ -20,27 +20,31 @@
...
+  text-align: right;
+  padding-right: 2em;
+  color: green;
```
Comment (NathanielScottStevens) [inline apps/andi/assets/css/edit.scss]: Would it make sense to create a color variable and tie this to the checkmark color?

File: apps/andi/lib/andi_web/live/edit_live_view.ex [hunk 3]
```diff
@@ -2,89 +2,184 @@ defmodule AndiWeb.EditLiveView do
...
+              <%= error_tag(fp, :temporal) %>
+            </div>
+            <div class="metadata-form__organization">
+              <%= Form.label(fp, :orgTitle, "Organization", class: "label label--required") %>
+              <%= Form.text_input(fp, :orgTitle, [class: "input input--text", disabled: true]) %>
+              <%= error_tag(fp, :orgTitle) %>
+            </div>
+            <div class="metadata-form__language">
+              <%= Form.label(fp, :language, "Language", class: "label") %>
+              <%= Form.select(fp, :language, [[key: "English", value: "english"], [key: "Spanish", value: "spanish"]], value: get_language(Form.input_value(fp, :language)), class: "select") %>
```
Comment (NathanielScottStevens) [inline apps/andi/lib/andi_web/live/edit_live_view.ex]: I would pull the select options out into a module attribute

File: apps/andi/assets/css/edit.scss (hunk 1 above)
Comment (ManApart) [reply apps/andi/assets/css/edit.scss]: Lol, I thought YOU made that change. We'll call it something else.

File: apps/andi/assets/css/edit.scss (hunk 2 above)
Comment (ManApart) [reply apps/andi/assets/css/edit.scss]: I was going to wait until it was used in more than one place, but we can do that.

File: apps/andi/lib/andi_web/live/edit_live_view.ex (hunk 3 above)
Comment (ManApart) [reply apps/andi/lib/andi_web/live/edit_live_view.ex]: JMS

File: apps/andi/lib/andi_web/live/edit_live_view.ex [hunk 4]
```diff
@@ -2,89 +2,184 @@ defmodule AndiWeb.EditLiveView do
...
+
     """
   end
 
   def mount(%{dataset: dataset}, socket) do
-    {:ok, assign(socket, dataset: dataset)}
+    new_business = dataset.business |> Map.from_struct()
+    new_technical = dataset.technical |> Map.from_struct()
+
+    change =
```
Comment (NathanielScottStevens) [inline apps/andi/lib/andi_web/live/edit_live_view.ex]: I would put this pipe into `Andi.DatasetSchema`. Maybe into the changeset function or perhaps a different changeset function (having multiple special changeset functions seems to be idiomatic)

File: apps/andi/lib/andi_web/live/edit_live_view.ex [hunk 5]
```diff
@@ -2,89 +2,184 @@ defmodule AndiWeb.EditLiveView do
...
+      |> Map.put(:business, new_business)
+      |> Map.put(:technical, new_technical)
+      |> Andi.DatasetSchema.changeset()
+
+    {:ok, assign(socket, changeset: change, is_saved: false)}
   end
 
-  defp get_private(%{technical: %{private: true}}), do: "Private"
-  defp get_private(_), do: "Public"
+  def handle_event(
```
Comment (NathanielScottStevens) [inline apps/andi/lib/andi_web/live/edit_live_view.ex]: I think we've talked about this some before, but reading this function it _seems_ like you should get everything that happens here for free. At the very least I would think you could do a `Changeset.apply` and avoid the `put_in` pipe.

File: apps/andi/assets/css/edit.scss (hunk 2 above)
Comment (ManApart) [reply apps/andi/assets/css/edit.scss]: It was also used as a checkmark, as you had guessed, so we're updating this as well.

File: apps/andi/lib/andi_web/live/edit_live_view.ex (hunk 5 above)
Comment (ManApart) [reply apps/andi/lib/andi_web/live/edit_live_view.ex]: The `put_in` lines are parsing keywords and preventing changes to 'read-only' fields. I'll add a comment.

File: apps/andi/lib/andi_web/live/edit_live_view.ex [hunk 6]
```diff
@@ -2,89 +2,171 @@ defmodule AndiWeb.EditLiveView do
...
+        %{"dataset_schema" => dataset_schema} = event,
+        socket
+      ) do
+    {:noreply, assign(socket, changeset: apply_changes(dataset_schema), is_saved: false)}
+  end
+
+  def handle_event("save", %{"dataset_schema" => dataset_schema} = event, socket) do
+    change = apply_changes(dataset_schema)
+
+    case change.valid? do
```
Comment (ManApart) [inline apps/andi/lib/andi_web/live/edit_live_view.ex]: This certainly reads better.

File: apps/andi/lib/andi_web/live/edit_live_view.ex [hunk 7]
```diff
@@ -2,89 +2,171 @@ defmodule AndiWeb.EditLiveView do
...
+
+        %{
+          original_dataset
+          | business: Map.merge(Map.from_struct(original_dataset.business), Map.from_struct(schema.business)),
+            technical: Map.merge(Map.from_struct(original_dataset.technical), Map.from_struct(schema.technical))
+        }
+        |> SmartCity.Dataset.new()
+        |> send_dataset_update
+
+        {:noreply, assign(socket, changeset: change, is_saved: true, update_stepper_state: "meta-data-save")}
```
Comment (ManApart) [inline apps/andi/lib/andi_web/live/edit_live_view.ex]: What's the update stepper state?

File: apps/andi/lib/andi_web/live/edit_live_view.ex [hunk 8]
```diff
@@ -2,89 +2,171 @@ defmodule AndiWeb.EditLiveView do
...
+
+      false ->
+        {:noreply, assign(socket, changeset: change, is_saved: false, display_errors: true)}
```
Comment (ManApart) [inline apps/andi/lib/andi_web/live/edit_live_view.ex]: Do we still need the `display_errors` workaround?

**Comment by ManApart** [comment]
It's looking a lot cleaner with the latest changes...

File: apps/andi/lib/andi_web/live/edit_live_view.ex (hunk 7 above)
Comment (kcarra) [reply apps/andi/lib/andi_web/live/edit_live_view.ex]: The arbitrary property that we have to add to get live view to re-render the view if there is no diff... At some point, I'm assuming we well perform some sort of UI action on save

File: apps/andi/lib/andi_web/live/edit_live_view.ex (hunk 8 above)
Comment (kcarra) [reply apps/andi/lib/andi_web/live/edit_live_view.ex]: I spent a decent bit of time looking into this and wasn't able to find the proper way to get the view to re-render. It's almost like live view is expecting some redirect to happen when a form is submitted and don't re render the view

File: apps/andi/lib/andi_web/live/edit_live_view.ex [hunk 9]
```diff
@@ -153,7 +154,9 @@ defmodule AndiWeb.EditLiveView do
     Brook.Event.send(instance_name(), dataset_update(), :andi, dataset)
   end
 
-  defp send_dataset_update({:error, e}), do: {:error, e}
+  defp send_dataset_update({:error, e}) do
+    Logger.warn("Unable to update dataset: #{IO.inspect({:error, e})}")
```
Comment (ManApart) [inline apps/andi/lib/andi_web/live/edit_live_view.ex]: IO.inspect?

File: apps/andi/lib/andi_web/live/edit_live_view.ex [hunk 10]
```diff
@@ -2,89 +2,178 @@ defmodule AndiWeb.EditLiveView do
...
+  defp apply_changes(data) do
+    data
+    |> put_in(["business", "keywords"], get_keywords_as_list(data["business"]["keywords"]))
+    |> Andi.DatasetSchema.changeset()
+  end
+
+  defp send_dataset_update({:ok, dataset}) do
+    case DatasetValidator.validate(dataset) do
+      :valid -> Brook.Event.send(instance_name(), dataset_update(), :andi, dataset)
+      {:invalid, errors} -> Logger.warn("Invalid dataset: #{IO.inspect({:invalid, errors})}")
```
Comment (ManApart) [inline apps/andi/lib/andi_web/live/edit_live_view.ex]: Do we need an IO.inspect or can we just do inspect()
```suggestion
      {:invalid, errors} -> Logger.warn("Invalid dataset: #{inspect({:invalid, errors})}")
```

Same question for a few lines down.

**Pull request merged**